File: C++/20ValidParentheses.cpp
//
//  20ValidParentheses.cpp
//  Testing C++
//
//  Created by Cem Beyenal on 1/9/25.
//
#include <iostream>  // For input/output (if needed for testing)
#include <string>    // For std::string
#include <stack>     // For std::stack

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stck;
        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{')
            {
                stck.push(s[i]);
            }
            else if (stck.empty())
            {
                return false;
            }
            else if (s[i] == ')' && stck.top() == '(')
            {
                stck.pop();
            }
            else if (s[i] == ']' && stck.top() == '[')
            {
                stck.pop();
            }
            else if (s[i] == '}' && stck.top() == '{')
            {
                stck.pop();
            }
            else
            {
                return false;
            }
        }
        return stck.empty();
    }
};

File: Python/20ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            elif not stack:
                return False
            elif stack[-1] == '(' and c == ')':
                stack.pop()
            elif stack[-1] == '[' and c == ']':
                stack.pop()
            elif stack[-1] == '{' and c == '}':
                stack.pop()
            else:
                return False
        return len(stack) == 0